### ISSUE TITLE
Node spacing/collision with each other

### ISSUE THREAD

**Issue by duhmojo** [vasturiano/react-force-graph#337]
Hi. I'm working with a React component and I'm trying to resize and space my nodes out naturally. There's overlapping now and I need to push them out they they can't overlap. I'm not trying to simulate anything, I just want the force direct layout to ensure that a group of nodes connected to another central node don't cluster together so they can't be viewed. I want to ensure that each node in a 2D graph is represented clearly, even if it means the view has to zoom out to see it all. 

I looked at the collision demo but it's not clear to me what's going on here. Variables like N and nodes aren't obvious. I'm also working with a React component class so there's some refactoring going on in my head. 
 
https://github.com/vasturiano/react-force-graph/blob/master/example/collision-detection/index.html

How can I adjust/customize node bounding boxes? Any help is appreciated. Thanks.

As a bonus question, for layouts like left-to-right, when a node is not connected (like say I have a tree and 2 disconnected nodes) the vertical spacing between the disconnected nodes is very wide/tall, and the connected nodes are bundled over top of each other, very tight. How do I adjust the spacing between nodes altogether? (I think this might be tided up/related to my initial question). Thanks.

**Comment by vasturiano** [comment]
@duhmojo thanks for reaching out.

I'd recommend looking over the docs for the [d3-force collision force plugin](https://github.com/d3/d3-force#collision). It hopefully provides some insight into how to configure these additional forces in a way that it suits your case.

As for your additional question, that may be due to the intensity of the existing link and charge (repelling) forces. Making the force engine work for specific cases requires calibrating the intensity of these forces so that they balance each other in an intuitive way. You can look over their docs which include all the parameters you can experiment with in the calibration:
* https://github.com/d3/d3-force#links
* https://github.com/d3/d3-force#many-body

**Comment by duhmojo** [comment]
Thanks Vasturiano. I've figured out what I was doing wrong in accessing the d3 force plugin. I wasn't explicitly installing the d3 library and `import * as d3 from "d3"`. In your example, because you link externally to D3 I didn't make the connection.

I'm was still struggling with accessing d3Force from a ForceGraph2D ref. I ended up getting close by sticking all my d3Force configuration into the ForceGraph2D ref as a function. No matter what I tried, the graph ref was always null, even in componentWillMount, so I had to put all my d3Force code into the ref= function. This might be something related to how ForceGraph2D renders, or when it renders, etc...

Here's what I'm attempting:

```
render() {
   return (
      <ForceGraph2D
         ref={(e) => {
            e.d3Force('center', null)
            e.d3Force('charge', null)
            const collide = d3.forceCollide(80)
            collide.strength = 1
            collide.radius = 80
            e.d3Force('collide', collide)

            this.graphCanvas = e
   )
}
```

I know the ref code and being executed but I see no difference in the presented graph. Depending on the layout, I have over lapping nodes, and when I drag them near each other its clear it's just the centre of the node that's avoiding collision.

Any tips are greatly appreciated.

**Comment by vasturiano** [comment]
I'm not familiar with that form of accessing the component's ref. Normally I see it with `useRef`, but perhaps it boils down to the same, and it's just a syntactic difference. 

In any case if the code runs, and only once after the initial render, maybe the issue is in the force config itself. You can try exaggerating the node collision radius to a higher number to see if you notice a change.

**Comment by duhmojo** [comment]
I would normally just `this.graphCanvas = React.createRef()` in the constructor and then `ref={ this.graphCanvas }` in ForceGraph2D. Then in `componentWillMount` I should have access to the ref because it fires once render is complete, but in this case ForceGraph2D ref doesn't seem to be set at all when `componentWillMount` fires. Its either something I'm doing wrong or something in how ForceGraph2D renders (probably my fault). I was able to play with d3Force by configuring it in the `ref={(e) => { ... e.d3Force() } }`.

What I'm wondering about is your collision code vs something like `fg.d3Force('collision', d3.forceCollide(10))`. The bounding box part of your example basically assigns a vx,vy to each node. I just want to make sure that nodes don't overlap.

With forceCollide I find that a basic `radial in` will keep things ok, but I think that's only relative to the number of nodes and their links. When I switch to a top-bottom or left-right the links and directional force supersede the forceCollide strength, and it seems to be situational. More nodes with concentrated links will force nodes together and if there are enough, it'll over whelm the forceCollide strength.

![image](https://user-images.githubusercontent.com/1214395/148316992-0e889b47-7e93-4912-91f4-4409182ff365.png)

![image](https://user-images.githubusercontent.com/1214395/148317026-5c388096-06c7-4a20-be2c-a352bfd595a1.png)


Any suggestions?

**Comment by duhmojo** [comment]
If I try beefing the collision strength to 30 in the left/right layout I get an unbalanced spacing. I think I need to work out how to set node boundaries.

![image](https://user-images.githubusercontent.com/1214395/148317705-e2fa4976-3130-4dde-a71e-5df390bf3557.png)

**Comment by vasturiano** [comment]
The bounding box force seen in the example is a totally separate force plugin from the collision force. 
https://github.com/vasturiano/react-force-graph/blob/c2805674e0528a80ff2478439372a71b9b4113c7/example/collision-detection/index.html#L35-L45

The intent of this force is to create container walls where the nodes bounce off of. It doesn't sound like that's at all relevant for your use case, so you don't need to add it. The only statement that relates to the node collision is this one, where a collision node radius of `4` is set:
https://github.com/vasturiano/react-force-graph/blob/c2805674e0528a80ff2478439372a71b9b4113c7/example/collision-detection/index.html#L34

**Comment by carljoachim** [comment]
@vasturiano Maybe a new issue, but I do get a TypeScript error when trying to use d3.forceCollide() in the d3Force method. Doesn't seems like d3Force accept a d3-method directly?

```
TS2345: Argument of type 'ForceCollide<NodeObject>' is not assignable to parameter of type 'ForceFn'.
  Types of property 'initialize' are incompatible.
    Type '(nodes: NodeObject[], random: () => number) => void' is not assignable to type '(nodes: NodeObject[]) => void'.
          fg.d3Force('collide', d3.forceCollide(50));
                                  ^^^^^^^^^^^^^^^^^^^
 | }
 }, []);

```

**Comment by vasturiano** [comment]
@carljoachim thanks for reporting. These types have now been fixed in the latest version.
